TM1py/Services/CellService.py

```diff
@@ -368,6 +370,57 @@ def write_dataframe(self, cube_name: str, data: 'pd.DataFrame', dimensions: Iter
                           use_changeset=use_changeset,
                           **kwargs)
 
+    @require_pandas
+    @manage_transaction_log
+    def write_dataframe_async(self, cube_name: str, data: 'pd.DataFrame',
+                              slice_size_of_dataframe: int, max_workers: int,
```

Comment:
`max_workers` can be made a default argument.

Reply:
Hi, @rkvinoth 
Thanks. The reason I don't want to put default values for the slice_size_of_dataframe and max_workers is that they are really dependent on the actual running environment. e.g. memory, cpu, network etc. After testing, I reckon it would be good to let users decide to get the most optimum combination. Obviously, if putting the too many workers, it would cause high cpu usage and freeze the system possibly.

Reply:
Fair enough. `max_workers` it totally depending on the environment.

However, for the `slice_size` I do believe we should provide a sensible default value since a casual TM1py user may not be aware of the fact that TM1 consumes records in chunks of 32k (since this is the max number of CellPut statements we can squeeze into an unbound process)

Reply:
`max_workers` by default takes the best possible number of threads. If users need optimization (or reach the threshold) they can reduce it as required.
`ThreadPoolExecutor` has the `max_workers` as a default argument and I think we should respect that.

with <3.8 the default count is cpu_count * 5
with >3.8 the default count is cpu_count + 4

Reply:
It is not guaranteed that Python runs on the same machine as TM1. 
TM1py can not know for sure how many CPUs are available to TM1.

Reply:
> It is not guaranteed that Python runs on the same machine as TM1.

If python is not running on the same machine as TM1, we shouldn't even worry about the number of threads.

> TM1py can not know for sure how many CPUs are available to TM1.

TM1py is just triggering the ThreadPoolExecutor.... ThreadPoolExecutor in turn uses `os` module to handle the thread count when it is not provided.
If you think ThreadPoolExecutor could take up more threads and not provide enough for TM1, I would argue that users could do the same. They can just go berserk and it is their responsibility to keep things in check.

Reply:
> If python is not running on the same machine as TM1, we shouldn't even worry about the number of threads.

Thanks for your thoughts and feedback @rkvinoth. I think I disagree. As TM1 admins or developers, I think we do care about not overloading the TM1 server with work.
I mean when TM1py sends those CellPut statements to the TM1 Server, the heavy lifting is done by TM1. Python is actually just waiting for a response from the server most of the time. TM1 is digesting the values and updating the cube structures. 
On the client-side python can use 4 or 10 threads even if you have just one free CPU. That's not problematic because most of the (python) threads will be waiting most of the time anyways for TM1 to finish digesting the API requests.
 
On the server-side, I would expect to see (in Arc, Pulse or tm1top) 4 or 10 threads processing almost at every moment.
So on the server side, we are causing TM1 to be busy. If the server has let's say 16 CPUs, I would argue we shouldn't use all of them for writing because then if TM1 users open views at this moment there are limited CPU resources to cater for their session and they may experience delays.

Reply:
I agree with what you have in mind on not burdening TM1 server. My point is the same, TM1 developer/admin should know what he is doing.
If you think developers will not provide the thread count if we leave it as default then let's not keep it default. I wanted to respect what `ThreadPoolExecutor` offers.

Reply:
As you know, using all the available threads will not provide better results. Developers still have to play with the thread count to arrive at the optimal number.

Reply:
Yeah maybe let's not provide a default for `max_workers`. It feels wrong to me to default to something that python suggests for the environment in which python runs. 
I think it's reasonable to ask the TM1py user to provide a value for `max_workers`.

> As you know, using all the available threads will not provide better results. Developers still have to play with the thread count to arrive at the optimal number.

Yes!

Reply:
Totally agree. I have same feeling when I was writing a huge dataframe on Cloud down to a TM1 server. I can simply put a very high number for the function to make the TM1 server "dead" as the CPU was exhausted.

So, do you guy want me to leave the max_workers without any default value as it is?

> Yeah maybe let's not provide a default for max_workers. It feels wrong to me to default to something that python suggests for the environment in which python runs.
I think it's reasonable to ask the TM1py user to provide a value for max_workers.

> It is not guaranteed that Python runs on the same machine as TM1.
TM1py can not know for sure how many CPUs are available to TM1.

Reply:
OK. So we don't put a default value for `max_workers`.